File: src/main/java/fi/metatavu/metaform/server/drafts/DraftController.java
package fi.metatavu.metaform.server.drafts;

import java.util.List;
import java.util.Map;
import java.util.UUID;

import javax.enterprise.context.ApplicationScoped;
import javax.inject.Inject;

import org.apache.commons.lang3.ArrayUtils;
import org.apache.commons.lang3.StringUtils;

import fi.metatavu.metaform.server.persistence.dao.DraftDAO;
import fi.metatavu.metaform.server.persistence.model.Draft;
import fi.metatavu.metaform.server.persistence.model.Metaform;

import com.fasterxml.jackson.databind.ObjectMapper;

import org.slf4j.Logger;

/**
 * Controller for draft related operations
 * 
 * @author Antti Leppä
 */
@ApplicationScoped
public class DraftController {
  
  @Inject
  private Logger logger;

  @Inject
  private DraftDAO draftDAO;
  
  /**
   * Creates new draft
   * 
   * @param metaform metaform
   * @param userId user id
   * @param data draft data
   * @return created draft
   */
  public Draft createDraft(Metaform metaform, UUID userId, Map<String, Object> data) {
    return draftDAO.create(UUID.randomUUID(), userId, metaform, serializeData(data));
  }

  /**
   * Finds a draft by id
   * 
   * @param id id
   * @return found draft or null if not found
   */
  public Draft findDraftById(UUID id) {
    return draftDAO.findById(id);
  }

  /**
   * Updates a draft
   * 
   * @param draft draftt
   * @param data data
   * @return updated draft
   */
  public Draft updateDraft(Draft draft, Map<String, Object> data) {
    return draftDAO.updateData(draft, serializeData(data));
  }

  /**
   * Deletes an draft
   * 
   * @param draft draft to be deleted
   */
  public void deleteDraft(Draft draft) {
    draftDAO.delete(draft);
  }

  /**
   * Serializes data as string
   * 
   * @param data data
   * @return data as string
   */
  private String serializeData(Map<String, Object> data) {
    try {
      ObjectMapper objectMapper = new ObjectMapper();
      return objectMapper.writeValueAsString(data);
    } catch (Exception e) {
      logger.error("Failed to serialize draft data", e);
    }
    
    return null;
  }
  
}

File: src/main/kotlin/fi/metatavu/metaform/server/controllers/DraftController.kt
package fi.metatavu.metaform.server.controllers

import com.fasterxml.jackson.databind.ObjectMapper
import fi.metatavu.metaform.server.exceptions.MalformedDraftDataException
import fi.metatavu.metaform.server.persistence.dao.DraftDAO
import fi.metatavu.metaform.server.persistence.model.Draft
import fi.metatavu.metaform.server.persistence.model.Metaform
import org.slf4j.Logger
import java.util.*
import jakarta.enterprise.context.ApplicationScoped
import jakarta.inject.Inject


/**
 * Controller for Drafts
 */
@ApplicationScoped
class DraftController {

    @Inject
    lateinit var logger: Logger

    @Inject
    lateinit var draftDAO: DraftDAO

    /**
     * Creates new draft
     *
     * @param metaform metaform
     * @param userId user id
     * @param data draft data
     * @return created draft
     */
    fun createDraft(
            metaform: Metaform,
            userId: UUID,
            data: String
    ): Draft {
        return draftDAO.create(
                id = UUID.randomUUID(),
                userId = userId,
                metaform = metaform,
                data = data,
        )
    }

    /**
     * Finds a draft by id
     *
     * @param id id
     * @return found draft or null if not found
     */
    fun findDraftById(id: UUID): Draft? {
        return draftDAO.findById(id)
    }

    /**
     * Updates a draft
     *
     * @param draft draft
     * @param data data
     * @return updated draft
     */
    fun updateDraft(draft: Draft, data: String): Draft {
        return draftDAO.updateData(draft, data)
    }

    /**
     * Deletes an draft
     *
     * @param draft draft to be deleted
     */
    fun deleteDraft(draft: Draft) {
        draftDAO.delete(draft)
    }

    /**
     * Lists drafts by metaform
     *
     * @param metaform metaform
     * @param firstResult first result
     * @param maxResults max results
     */
    fun listByMetaform(metaform: Metaform, firstResult: Int?, maxResults: Int?): List<Draft> {
        return draftDAO.listByMetaform(metaform, firstResult, maxResults)
    }

    /**
     * Serializes data as string
     *
     * @param data data
     * @return data as string
     */
    @Throws(MalformedDraftDataException::class)
    fun serializeData(data: Map<String, Any>): String {
        try {
            val objectMapper = ObjectMapper()
            return objectMapper.writeValueAsString(data)
        } catch (e: Exception) {
            throw MalformedDraftDataException("Failed to serialize draft data", e)
        }
    }
}